Fix issues with `QuadBakingVertexConsumer` and Forge's OBJ loader

SuperMartijn642 opened the pull request:
As described in #10064, `QuadBakingVertexConsumer` currently results in an exception when submitting 4 vertices.

This is caused by `QuadBakingVertexConsumer#addVertex` calling `QuadBakingVertexConsumer#endVertex` which immediately increases the vertex index, hence vertices are written starting at vertex index 1 rather than 0.
To account for this, I added a `started` flag which is initially false and updated to `true` when starting the first vertex. Only when a vertex is started and the `started` flag is already true, will the vertex index now be increased.

As vertices are no longer ended explicitly, but rather when a new vertex is started, this would mean the last vertex is never ended. To solve this, I changed `QuadBakingVertexConsumer` from having a consumer for quads to having a `#finishQuad` method which simply creates the quad and returns it.

With these changes, the example from #10064 now loads without errors:
<img src="https://github.com/user-attachments/assets/451037b4-fdb9-4407-8ed3-e51db7d0674f" width="500" />
(Yes, that's how it is supposed to look)

This PR does break pretty much any mod which interacts with the `QuadBakingVertexConsumer` class, but I don't know if there's a good solution without breaking changes to `QuadBakingVertexConsumer`.

LexManos commented:
Ya its a simple issue, would be nice to have a test mod that uses OBJ loading. 
I'm not a fan of the whole OBJ/Rendering system in Forge, but I don't think binary breaks like this is a good idea right now.
Wouldn't a simpler solution be to just set the default value of vertexIndex to -1 so that the first time addVertex is called it gets set to 0?
Tho that doesn't solve the 'last vertex is never ended' issue... humm

If you're available tho, I would like to see a simple test mod that adds a single block/item with a OBJ model. Just to make sure we have tests for the future.

Edit: I have some ideas on how to clean this code up. Can write them up if there is a test mod.

A simple change that could work, still required manually calling bake() but so does yours, but this variant doesn't have binary breaks. If I were to re-write this I would skip the 'consumer' crap and just make this a simple 'BakedQuadBuilder'. 

```diff
     private final Map<VertexFormatElement, Integer> ELEMENT_OFFSETS = Util.make(new IdentityHashMap<>(), map -> {
         int i = 0;
         for (var element : DefaultVertexFormat.BLOCK.getElements())
             map.put(element, DefaultVertexFormat.BLOCK.getOffset(i++) / 4); // Int offset
     });
-    private static final int QUAD_DATA_SIZE = IQuadTransformer.STRIDE * 4;
+    private static final int MAX_VERTICES = 4;
+    private static final int QUAD_DATA_SIZE = IQuadTransformer.STRIDE * MAX_VERTICES;
 
     private final Consumer<BakedQuad> quadConsumer;
 
-    int vertexIndex = 0;
+    int vertexIndex = -1;
     private int[] quadData = new int[QUAD_DATA_SIZE];
 
     private int tintIndex;
```
```diff
     @Override
-    public VertexConsumer addVertex(float x, float y, float z)
-    {
-        endVertex();
+    public VertexConsumer addVertex(float x, float y, float z) {
+        if (vertexIndex++ == MAX_VERTICES)
+            build();
 
         int offset = vertexIndex * IQuadTransformer.STRIDE + IQuadTransformer.POSITION;
         quadData[offset] = Float.floatToRawIntBits((float) x);

```
```diff
-    private void endVertex()
-    {
-        if (vertexIndex != 4) {
-            ++vertexIndex;
-            return;
-        }
+    public BakedQuad build() {
+        var quad = new BakedQuad(quadData, tintIndex, direction, sprite, shade, hasAmbientOcclusion);
         // We have a full quad, pass it to the consumer and reset
-        quadConsumer.accept(new BakedQuad(quadData, tintIndex, direction, sprite, shade, hasAmbientOcclusion));
+        quadConsumer.accept(quad);
         vertexIndex = 0;
         quadData = new int[QUAD_DATA_SIZE];
+        return quad;
     }
```

SuperMartijn642 committed:
Add an OBJ model test

SuperMartijn642 committed:
Fix issues with `QuadBakingVertexConsumer`

SuperMartijn642 commented:
Implemented the fix as you suggested. There shouldn't be any binary breaks, only the need to now call `#build` after the final quad.

I also added a simple test which renders an obj model at 0,0,0 in the world. (I created the model using BlockBench, so shouldn't be any copyright issues)
![image](https://github.com/user-attachments/assets/14487b1a-8568-42ac-a2a4-1d7a79078044)
I am not completely sure how to properly add a test and mostly just copied how other test classes did things. If something should be done differently, let me know.

LexManos commented:
Seems to work, I would like it better if we had a more populated suite of OBJ files to load. And actually test the data is loaded how we expect it to be so that it could be turned into a GameTest. As well as just re-writing this entire API so that its not consumer hell. 

But this is good for now.
